marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowMigrator.java

```diff
@@ -0,0 +1,191 @@
+                        Path targetDir = step.getStepDefinitionType().equals(StepDefinition.StepDefinitionType.INGESTION) ? ingestionDir : mappingDir;
+                        String stepFileName = new StringBuilder(step.getName()).append(".step.json").toString();
+                        File stepFile = targetDir.resolve(stepFileName).toFile();
+                        logger.info("Creating step artifact "+ stepFile.toString());
+                        if (stepFile.exists()) {
+                            String msg = "Step artifact " + stepFile.toString() + " already exists. The step artifact will be written to ";
+                            //Update step artifact with new name
+                            String stepName = new StringBuilder(flow.getName()).append("-").append(step.getName()).toString();
+                            newStepArtifact.put("name", stepName);
+                            //Update the filename
+                            stepFileName = new StringBuilder(flow.getName()).append("-").append(stepFileName).toString();
+                            stepFile = targetDir.resolve(stepFileName).toFile();
+
+                            logger.info(msg + stepFile.toString()) ;
+                        }
+                        try{
+                            writer.writeValue(stepFile, newStepArtifact);
+                            logger.info("Step artifact " + stepFile.toString() + " successfully created.");
+                        }
+                        catch(IOException e){
+                            logger.error("Step artifact " + stepFile.toString() + " creation failed: " + e.getMessage());
+                        }
+                    }
+                    else {
+                        logger.info("Step " + step.getName() + " is not an out of the box ingestion or mapping step. It will remain inline inside the flow artifact");
+                        newSteps.set(entry.getKey(), mapper.valueToTree(step));
+                    }
+                }
+                newFlow.set("steps",newSteps);
+                File flowFile = Paths.get(hubProject.getFlowsDir().toString(), flow.getName() + FlowManager.FLOW_FILE_EXTENSION).toFile();
+                try{
+                    writer.writeValue(flowFile, newFlow);
+                    logger.info("Flow " + flowFile.toString() + " successfully migrated.");
+                }
+                catch(IOException e){
+                    logger.error("Flow artifact " + flowFile.toString() + " creation failed: " + e.getMessage());
+                }
+            }
+        });
+    }
+
+    // Only create step artifacts for ootb ingestion and mapping steps. Other steps(including custom ingestion and mapping)
+    // will be inline
+    private boolean stepRequiresMigration(Step step) {
+        return (StepDefinition.StepDefinitionType.MAPPING.equals(step.getStepDefinitionType()) && "entity-services-mapping".equalsIgnoreCase(step.getStepDefinitionName())) ||
```

Comment:
endpoint http://localhost:8080/api/artifacts/mapping seems to return all the versions of a mapping. Shouldnt it return only the latest version. UI ends up showing all the versions which is not desired.
<img width="546" alt="image" src="https://user-images.githubusercontent.com/16124406/82189188-da6eb100-98a3-11ea-9d2c-94fe8e4d3c50.png">

Reply:
It will return whatever version is specified in the flow. Also, "mapping" object will be removed from the mapping step